### PR TITLE
Cornami, Inc : contribute CMake build files

### PR DESCRIPTION
**Pull request by ai-mannamalai**
Provide CMakeLists.txt files to build all of the core components of HEIR Project (except Yosys and GTest bits)
Contributed by Cornami, Inc, www.cornami.com 
Two ways to build the code:

## Using LLVM build folder
```bash
$ cmake -B builddir -S ./ -DMLIR_DIR=/path/to/llvm-project/build/lib/cmake/mlir -DLLVM_DIR=/path/to/llvm-project/build/lib/cmake/llvm && cmake --build builddir
```

## Using Installed LLVM / MLIR
```bash
$ cmake -S ./ -B builddir -DENABLE_INSTALLED_MLIR=True -DCMAKE_PREFIX_PATH=/usr/local/lib/cmake -DMLIR_CMAKE_DIR=/usr/local/lib/cmake/mlir/ -DLLVM_CMAKE_DIR=/usr/local/lib/cmake/llvm/ &&  cmake --build builddir
```

### CODE REVIEW

File: CMakeLists.txt
Comment (j2kun) [inline CMakeLists.txt:16]:
```suggestion
option(ENABLE_YOSYS "Enable YOSYS for BFV and Comb dialects (optional)" FALSE)
```
```suggestion
option(ENABLE_YOSYS "Enable YOSYS for yosys-optimizer pass (optional)" FALSE)
```

File: lib/Analysis/SelectVariableNames/CMakeLists.txt
Comment (j2kun) [inline lib/Analysis/SelectVariableNames/CMakeLists.txt:7]: Looks like a mix of tabs, spaces, and the space-indented parts don't all have the same space indentation. Can you `expandtab`? Or maybe something like

```bash
sed -i 's/^\s\{2,\}/    /g' **/CMakeLists.txt
```

**Comment by j2kun** [comment]
Great work! I think we should get a quick agreement about the expectations for maintaining this build going forward. Some questions:

- Does the CMake build have to pass CI to submit a PR? If so, we should add the CI workflow in this or the next PR. If not, do we expect CMake users to periodically fix a broken CMake build (this is what upstream LLVM does for the bazel build).
- Are we expecting the user to bring their own LLVM repo and set up the path? If so, we should document how to do that somewhere, ideally so it shows up on `heir.dev`, but for now having the CI workflow document it would suffice.
- How will we keep the LLVM commit in sync between the bazel and CMake builds? At least in the CI, can we parse `bazel/import_llvm.bzl` for the commit hash for this?

**Comment by asraa** [comment]
> Does the CMake build have to pass CI to submit a PR? If so, we should add the CI workflow in this or the next PR. If not, do we expect CMake users to periodically fix a broken CMake build (this is what upstream LLVM does for the bazel build).

I was about to broach an issue to talk about this - I would like to see a CI workflow added (probably in a follow-up, because I expect some non-trivial work to handle pulling in and caching LLVM). I would hope right now that it's non-blocking, and that users who are interested in the CMake proactively fix it (we can get the CI to auto-file an issue saying CMake is broken and tag "CMake build maintainers").

> How will we keep the LLVM commit in sync between the bazel and CMake builds? At least in the CI, can we parse bazel/import_llvm.bzl for the commit hash for this?

+1

File: lib/Analysis/SelectVariableNames/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Analysis/SelectVariableNames/CMakeLists.txt]: I something similar - fixed up now.

**Comment by ai-mannamalai** [comment]
> > Does the CMake build have to pass CI to submit a PR? If so, we should add the CI workflow in this or the next PR. If not, do we expect CMake users to periodically fix a broken CMake build (this is what upstream LLVM does for the bazel build).
> 
> I was about to broach an issue to talk about this - I would like to see a CI workflow added (probably in a follow-up, because I expect some non-trivial work to handle pulling in and caching LLVM). I would hope right now that it's non-blocking, and that users who are interested in the CMake proactively fix it (we can get the CI to auto-file an issue saying CMake is broken and tag "CMake build maintainers").
> 
> > How will we keep the LLVM commit in sync between the bazel and CMake builds? At least in the CI, can we parse bazel/import_llvm.bzl for the commit hash for this?
> 
> +1

- We should likely add CMake build in the minimum for PRs to pass otherwise things will diverge quickly
- I'm not expert in building LLVM quickly but I assume other projects downstream of MLIR will be testing in similar way with CMake and we can reuse their methods for CI
- Currently this CMakeLists.txt assumes the mlir/include/ and llvm/include/ folders to exist - this needs some TLC
- Currently the heir-opt and heir-lsp do not built and we have to fix the link errors stopping this build

**Comment by ai-mannamalai** [comment]
Updated pre-commit checks as per docs https://heir.dev/docs/getting_started/

File: lib/Transforms/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Transforms/CMakeLists.txt:11]: Can you create an issue for this and link to it here + anywhere else changes would have to be made? 
(there's a CI thing that keeps track of mentions for each issue)
```suggestion
    # TODO (#999): Support Yosys in cmake build.
    message(FATAL_ERROR "Yosys transforms not supported in CMake build yet; use Bazel instead.")
```

File: lib/Analysis/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Analysis/CMakeLists.txt:1]: It seems like `MulDepthAnalysis` and `SecretnessAnalysis` don't yet have a CMakeLists.txt (they were recently added)
```suggestion
add_subdirectory(MulDepthAnalysis)
add_subdirectory(RotationAnalysis)
add_subdirectory(SecretnessAnalysis)
```

File: lib/Dialect/ArithExt/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Dialect/ArithExt/CMakeLists.txt:2]: It would be nice if the prefix was consistent across all CMake targets, currently it seems like some are prefixed with `MLIR` and some with `MLIRHEIR`. Maybe we just want to go for `HEIR`?

```suggestion
add_mlir_dialect_library(HEIRArithExt
```

File: lib/Dialect/BGV/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Dialect/BGV/CMakeLists.txt:5]: I'm not sure if this is an issue, but in Bazel, the transforms are in their own target (and therefore library), whereas in the current cmake, they're being combined into a single target.
One downside to this that I could forsee is keeping dependency lists in sync between Bazel and CMake, as CMake would have some of the dependencies "hidden".

File: lib/Analysis/RotationAnalysis/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Analysis/RotationAnalysis/CMakeLists.txt:1]: I'm not sure if `add_mlir_dialect_library` is the right thing to use here, given that this is an analysis pass? 
According to https://mlir.llvm.org/docs/Tutorials/CreatingADialect/ it should be `add_mlir_conversion_library`?

Also, I think it'd be nice to keep the names of targets consistent across Bazel and CMake (modulo some prefix for CMake), e.g., `HEIRRotationAnalysis` for this target.

File: tools/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline tools/CMakeLists.txt:1]: There's a lot going on here and I'm not quite sure what's happening - if I compare this to `mlir-opt` upstream (https://github.com/llvm/llvm-project/blob/main/mlir/tools/mlir-opt/CMakeLists.txt) or to my own old `heco`-opt (https://github.com/MarbleHE/HECO/blob/main/CMakeLists.txt) they seem to be a lot simpler, but maybe you tried that approach and ran into some issues?

**Review by AlexanderViand-Intel — CHANGES_REQUESTED** [review]
Thanks so much for getting the ball rolling on CMake support for heir! While I've grown to appreciate bazel (although that might just be the stockholm syndrome talking 😉 ) I still remember how badly I missed cmake at the beginning. 

I expect this will make it quite a bit easier for people familiar with MLIR/C++ but not bazel to get started working with and on HEIR :) It might also make upstreaming features/reproducing bugs upstream/etc a bit easier, since there's already a working cmakelists to work from.

I spotted a few issues in my preliminary review and was a bit confused by the /tools CMakeLists.txt, so instead of trying to debug the heir-lsp issues directly, I figured I'd post my comments as-is - with a bit of luck, the lsp thing will sort itself out as part of the general cleanup :)

File: lib/Dialect/BGV/CMakeLists.txt
Comment (AlexanderViand-Intel) [reply lib/Dialect/BGV/CMakeLists.txt:5]: Oh, and I'm also wondering if there's anyway we can use `add_mlir_dialect` for the dialects themseves, as thathandles a LOT of the tablegen/etc stuff internally. See https://mlir.llvm.org/docs/Tutorials/CreatingADialect/#tablegen-targets and https://github.com/llvm/llvm-project/blob/main/mlir/cmake/modules/AddMLIR.cmake#L162 for what it does. I _think_ it continues gracefully if any of the files are missing (at least, I used it in HECO for dialects that only had a single *.td).

Alternatively, maybe we can define our own add_heir_dialect that works with our, slightly different, directory structure. My hope is that the simpler we make each CMakeLists.txt file, the more likely we'll be able to keep everything nicely in sync between Bazel and CMake

**Comment by AlexanderViand-Intel** [comment]
> * Does the CMake build have to pass CI to submit a PR? If so, we should add the CI workflow in this or the next PR. If not, do we expect CMake users to periodically fix a broken CMake build (this is what upstream LLVM does for the bazel build).

I think it'd be nice to add this to the CI (and PR requirements). If we extend the /scripts/template.py generator to emit the basic CMakeLists.txt things, too, the overhead should be pretty low for most people.  If something doesn't work "out of the box", I'd be happy to help fix up any PRs with failing CMake.

> * Are we expecting the user to bring their own LLVM repo and set up the path? If so, we should document how to do that somewhere, ideally so it shows up on `heir.dev`, but for now having the CI workflow document it would suffice.
> * How will we keep the LLVM commit in sync between the bazel and CMake builds? At least in the CI, can we parse `bazel/import_llvm.bzl` for the commit hash for this?

I think we could address both of these at the same time by using `ExternalProject_Add` to download and build LLVM via CMake, similar to what bazel does. It should be possible to write a bit of cmake magic to grab the tag from the bazel import_llvm.bzl file, though I probably won't get around to playing around with that until after the WAHC submission deadline (07/26)

File: lib/Dialect/ArithExt/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Dialect/ArithExt/CMakeLists.txt:2]: I'd just go with what BUILD / Bazel is using so we are consistent with the products of the build

File: lib/Analysis/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Analysis/CMakeLists.txt:1]: ok added it

File: lib/Dialect/ArithExt/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Dialect/ArithExt/CMakeLists.txt:2]: resolved in PR updated on Jul 17th, 11PM (PST)

File: lib/Dialect/BGV/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Dialect/BGV/CMakeLists.txt:5]: we had this conversation on Discord; you mentioned you didn't want to touch source files nor the BUILD Bazel files. Indeed my original implementation used the add_mlir_dialect macros in CMake but I had to remove them in place of piecemeal sub-macros to enable the builds to match the unique uses of the HEIR project.

As for the  sub targets in MLIRBGV I should go back and do what is done in BUILD folder; this horrible hack is a side-effect of rushing to get this to work

**Comment by ai-mannamalai** [comment]
Updated to build all the **heir-opt, heir-lsp** and **heir-translate** using CMakeLists.txt

**Comment by ai-mannamalai** [comment]
remaining item seems to be how to hook up all this into the CI system; honestly I'd like to tackle this in a second PR.
cc: @asraa @AlexanderViand-Intel

File: lib/Analysis/RotationAnalysis/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Analysis/RotationAnalysis/CMakeLists.txt]: TBH there is over 60+ files linked against the heir-opt binary; otherwise its plain and simple.

File: tools/CMakeLists.txt
Comment (ai-mannamalai) [reply tools/CMakeLists.txt:1]: there are 60+ libraries linked to build the heir-opt which makes this CMake file to be quite long; its comparable with the tools/BUILD file and not in upstream etc. apples :: apples comparison IMHO

File: lib/Analysis/RotationAnalysis/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Analysis/RotationAnalysis/CMakeLists.txt]: done

**Comment by ai-mannamalai** [comment]
> > * Does the CMake build have to pass CI to submit a PR? If so, we should add the CI workflow in this or the next PR. If not, do we expect CMake users to periodically fix a broken CMake build (this is what upstream LLVM does for the bazel build).
> 
> I think it'd be nice to add this to the CI (and PR requirements). If we extend the /scripts/template.py generator to emit the basic CMakeLists.txt things, too, the overhead should be pretty low for most people. If something doesn't work "out of the box", I'd be happy to help fix up any PRs with failing CMake.
> 
> > * Are we expecting the user to bring their own LLVM repo and set up the path? If so, we should document how to do that somewhere, ideally so it shows up on `heir.dev`, but for now having the CI workflow document it would suffice.
> > * How will we keep the LLVM commit in sync between the bazel and CMake builds? At least in the CI, can we parse `bazel/import_llvm.bzl` for the commit hash for this?
> 
> I think we could address both of these at the same time by using `ExternalProject_Add` to download and build LLVM via CMake, similar to what bazel does. It should be possible to write a bit of cmake magic to grab the tag from the bazel import_llvm.bzl file, though I probably won't get around to playing around with that until after the WAHC submission deadline (07/26)

this can be done in a followup PR IMO; issue #818

File: lib/Dialect/BGV/CMakeLists.txt
Comment (AlexanderViand-Intel) [reply lib/Dialect/BGV/CMakeLists.txt:5]: Let's merge this PR as-is, and I'll create an issue for me to wirte a custom set of `add_heir_{dialect/conversion/transform/analysis}` CMake macros. With those in place (and updating the right variables), we can then also simplify the tools build to be as simple as the upstream ones.

**Comment by AlexanderViand-Intel** [comment]
I'm getting the following errors during configuration (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`):

``` 
  Found erroneous configuration for source file Utils.cpp

  LLVM's build system enforces that all source files are added to a build
  target, that exactly one build target exists in each directory, and that
  this target lists all files in that directory.  If you want multiple
  targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
  specification, though it is discouraged.

  Please update /home/aviand/code/heir/lib/Conversion/CMakeLists.txt
  ```
```
Found erroneous configuration for source file UnusedMemRef.cpp

  LLVM's build system enforces that all source files are added to a build
  target, that exactly one build target exists in each directory, and that
  this target lists all files in that directory.  If you want multiple
  targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
  specification, though it is discouraged.

  Please update
  /home/aviand/code/heir/lib/Transforms/UnusedMemRef/CMakeLists.txt
```

```
  Found erroneous configuration for source file UnusedMemRef.cpp

  LLVM's build system enforces that all source files are added to a build
  target, that exactly one build target exists in each directory, and that
  this target lists all files in that directory.  If you want multiple
  targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
  specification, though it is discouraged.

  Please update
  /home/aviand/code/heir/lib/Transforms/UnusedMemRef/CMakeLists.txt
```

File: lib/Dialect/BGV/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Dialect/BGV/CMakeLists.txt:5]: sounds good to me - thanks @AlexanderViand-Intel

File: lib/Transforms/UnusedMemRef/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Transforms/UnusedMemRef/CMakeLists.txt:7]: mismatched case causes build issues on linux
```suggestion
    UnusedMemRef.cpp
```

File: lib/Conversion/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Conversion/CMakeLists.txt:3]:
```suggestion
    Utils.cpp
```

File: lib/Source/AutoHog/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Source/AutoHog/CMakeLists.txt:2]: Since the RapidJSON dependency probably isn't going to be updated frequently, I think it's save to already include it in this PR with a hardcoded git commit.

```suggestion
include(ExternalProject)

# Download RapidJSON
# TODO (#818): automatically extract the commit from the bazel WORKSPACE file 
ExternalProject_Add(
    rapidjson
    PREFIX "vendor/rapidjson"
    GIT_REPOSITORY "https://github.com/Tencent/rapidjson.git"
    GIT_TAG f54b0e47a08782a6131cc3d60f94d038fa6e0a51
    TIMEOUT 10
    CMAKE_ARGS
        -DRAPIDJSON_BUILD_TESTS=OFF
        -DRAPIDJSON_BUILD_DOC=OFF
        -DRAPIDJSON_BUILD_EXAMPLES=OFF
    CONFIGURE_COMMAND ""
    BUILD_COMMAND ""
    INSTALL_COMMAND ""
    UPDATE_COMMAND ""
)

# Prepare RapidJSON (RapidJSON is a header-only library)
ExternalProject_Get_Property(rapidjson source_dir)
set(RAPIDJSON_INCLUDE_DIR ${source_dir}/include)
```

**Review by AlexanderViand-Intel — CHANGES_REQUESTED** [review]
With your last update + these additional changes below, the project now configures succesfully for me :) 
 
However when I try to build (`cmake --build build --target heir-opt`), I get  an error that it cannot find `/mlir/include/mlir/IR/Builders.h`

```
fatal error: mlir/include/mlir/IR/Builders.h: No such file or directory
    4 | #include "mlir/include/mlir/IR/Builders.h"               // from @llvm-project
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by ai-mannamalai** [comment]
> With your last update + these additional changes below, the project now configures succesfully for me :)
> 
> However when I try to build (`cmake --build build --target heir-opt`), I get an error that it cannot find `/mlir/include/mlir/IR/Builders.h`
> 
> ```
> fatal error: mlir/include/mlir/IR/Builders.h: No such file or directory
>     4 | #include "mlir/include/mlir/IR/Builders.h"               // from @llvm-project
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> ```

I fixed this bug in the most recent push at 2:40PM PST.

**Comment by ai-mannamalai** [comment]
> I'm getting the following errors during configuration (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`):
> 
> ```
>   Found erroneous configuration for source file Utils.cpp
> 
>   LLVM's build system enforces that all source files are added to a build
>   target, that exactly one build target exists in each directory, and that
>   this target lists all files in that directory.  If you want multiple
>   targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
>   specification, though it is discouraged.
> 
>   Please update /home/aviand/code/heir/lib/Conversion/CMakeLists.txt
> ```
> 
>  ```
> Found erroneous configuration for source file UnusedMemRef.cpp
> 
>   LLVM's build system enforces that all source files are added to a build
>   target, that exactly one build target exists in each directory, and that
>   this target lists all files in that directory.  If you want multiple
>   targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
>   specification, though it is discouraged.
> 
>   Please update
>   /home/aviand/code/heir/lib/Transforms/UnusedMemRef/CMakeLists.txt
> ```
> 
>  ```
>   Found erroneous configuration for source file UnusedMemRef.cpp
> 
>   LLVM's build system enforces that all source files are added to a build
>   target, that exactly one build target exists in each directory, and that
>   this target lists all files in that directory.  If you want multiple
>   targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
>   specification, though it is discouraged.
> 
>   Please update
>   /home/aviand/code/heir/lib/Transforms/UnusedMemRef/CMakeLists.txt
> ```
> 
> LLVM commit is acc159aea1e641e3694ab8fe5faa231788077011 (current integrate for HEIR), CMake version is 3.28.3

resolved

File: lib/Dialect/CGGI/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Dialect/CGGI/CMakeLists.txt:18]:
```suggestion
        LINK_LIBS PUBLIC
        MLIRHEIRInterfaces
        MLIRLWE
        MLIRIR
```

File: lib/Dialect/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Dialect/CMakeLists.txt:32]:
```suggestion
    DEPENDS
    MLIRHEIRInterfacesOpsIncGen
```

File: lib/Source/AutoHog/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline lib/Source/AutoHog/CMakeLists.txt:30]:
```suggestion
    ADDITIONAL_HEADER_DIRS
    ${CMAKE_CURRENT_SOURCE_DIR}/
    
    DEPENDS
    rapidjson

```

File: CMakeLists.txt
Comment (AlexanderViand-Intel) [inline CMakeLists.txt:58]:
```suggestion

```

File: CMakeLists.txt
Comment (AlexanderViand-Intel) [inline CMakeLists.txt:62]: This little trick removes the need for the custom command (which unfortunately didn't work for me):
* first set of includes is for heir code (which uses a very long include path for some reason)
* second set is so that includes in included mlir/llvm code (shorter path) work, too
```suggestion
include_directories(${LLVM_INCLUDE_DIRS}/../..)
include_directories(${MLIR_INCLUDE_DIRS}/../..)
include_directories(${LLVM_INCLUDE_DIRS})
include_directories(${MLIR_INCLUDE_DIRS})
```

File: README.md
Comment (AlexanderViand-Intel) [inline README.md:41]: The other build documentation is in docs/content/docs/getting_started.md, and that's where I'd put the CMake documentation, too. However, I'd actually hold off on that a bit, until the CMake stuff is more mature.
```suggestion
```

File: tools/CMakeLists.txt
Comment (AlexanderViand-Intel) [inline tools/CMakeLists.txt:118]:
```suggestion
        MLIRMemrefToArithMemrefGlobalReplace
        MLIRMemRefToLLVM
        MLIRMemRefTransforms
```

**Review by AlexanderViand-Intel — CHANGES_REQUESTED** [review]
With the below changes, I can get it to configure (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`), and compile (`cmake --build build --target heir-opt -j`), and the same for `heir-lsp` and `heir-translate`  :)

File: CMakeLists.txt
Comment (ai-mannamalai) [reply CMakeLists.txt:49]: Okay - let me see

**Comment by ai-mannamalai** [comment]
> With the below changes, I can get it to configure (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`), and compile (`cmake --build build --target heir-opt -j`), and the same for `heir-lsp` and `heir-translate` :)

this seems to work on my Linux machine against a build LLVM folder; however it does not work against the installed LLVM / MLIR which is a kicker.

**Comment by AlexanderViand-Intel** [comment]
> > With the below changes, I can get it to configure (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`), and compile (`cmake --build build --target heir-opt -j`), and the same for `heir-lsp` and `heir-translate` :)
> 
> this seems to work on my Linux machine against a build LLVM folder; however it does not work against the installed LLVM / MLIR which is a kicker. The former code changes would do that. Let me see if I can resolve the two worlds

That's odd! Though I guess installed LLVM isn't really supportable long-term anyway, because of the LLVM integrate synchronization.

**Comment by ai-mannamalai** [comment]
> > > With the below changes, I can get it to configure (`cmake -B build -DMLIR_DIR=../llvm-project/build/lib/cmake/mlir`), and compile (`cmake --build build --target heir-opt -j`), and the same for `heir-lsp` and `heir-translate` :)
> > 
> > 
> > this seems to work on my Linux machine against a build LLVM folder; however it does not work against the installed LLVM / MLIR which is a kicker. The former code changes would do that. Let me see if I can resolve the two worlds
> 
> That's odd! Though I guess installed LLVM isn't really supportable long-term anyway, because of the LLVM integrate synchronization.

I introduced a optional compile pathway like this,
```bash
$ cmake -S ./ -B builddir -DENABLE_INSTALLED_MLIR=TRUE -DCMAKE_PREFIX_PATH=/usr/local/lib/cmake -DMLIR_CMAKE_DIR=/usr/local/lib/cmake/mlir/ -DLLVM_CMAKE_DIR=/usr/local/lib/cmake/llvm/  && cmake —build builddir -j 
```

the validation of both pathways is checked out on my Apple Mac OSX build
https://gist.github.com/ai-mannamalai/7d097bb610422889e8ec465b8ccea039

**Comment by ai-mannamalai** [comment]
If you're OK with the changes we can squash them and enable the tests etc.

**Comment by ai-mannamalai** [comment]
x-checked after final push works as expected with the optional flag.

**Comment by ai-mannamalai** [comment]
> x-checked after final push works as expected with the optional flag.

rebase as 1 commit

File: CMakeLists.txt
Comment (AlexanderViand-Intel) [inline CMakeLists.txt:77]: For the non-installed CMAKE option, both the "long" and "short" form of the include paths are needed, but otherwise everything seems to work 👍 
```suggestion
else()
include_directories(${LLVM_INCLUDE_DIRS})
include_directories(${MLIR_INCLUDE_DIRS})
include_directories(${LLVM_INCLUDE_DIRS}/../..)
include_directories(${MLIR_INCLUDE_DIRS}/../..)
endif()
```

**Review by AlexanderViand-Intel — APPROVED** [review]
Awesome!

**Comment by ai-mannamalai** [comment]
thanks @AlexanderViand-Intel - I x-checked as well and rebase to single commit;

**Review by AlexanderViand-Intel — APPROVED** [review]

**Comment by AlexanderViand-Intel** [comment]
> thanks @AlexanderViand-Intel - I x-checked as well and rebase to single commit;

@ai-mannamalai Did you push the latest rebase? I still see 2 commits in the PR :) 

Oddly enough, the "PR review status / require-single-commit-on-approval (pull_request_review)" check has nevertheless passed - @asraa: any idea what's going on there?

**Comment by ai-mannamalai** [comment]
@AlexanderViand-Intel - I did not rebase to single commit - you were right - forgot to push earlier. Now should be fixed up.

File: lib/Dialect/ArithExt/IR/CMakeLists.txt
Comment (j2kun) [inline lib/Dialect/ArithExt/IR/CMakeLists.txt:11]: We don't need any mlir docs, since CMake is not load-bearing for generating them. Let's remove all these calls.

**Review by j2kun — CHANGES_REQUESTED** [review]
I won't be able to provide much of a meaningful review, so it seems LGTM. CC @asraa to let here know this is a light review before being pull into Google

File: lib/Dialect/ArithExt/IR/CMakeLists.txt
Comment (ai-mannamalai) [reply lib/Dialect/ArithExt/IR/CMakeLists.txt:11]: ok

Comment (ai-mannamalai) [reply lib/Dialect/ArithExt/IR/CMakeLists.txt]: - remove add_mlir_doc as per review from @Jeremy Khun in latest push.

**Review by j2kun — APPROVED** [review]

**Comment by AlexanderViand-Intel** [comment]
Pinging @asraa :)

**Comment by asraa** [comment]
> Pinging @asraa :)

Sorry about that! @code-perspective was prepping the internal merge, I think it should be ready to merge soon

**Pull request merged**
